{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "eed9ce10-65b7-49bc-bb09-cd02497f827c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gen\tnevals\tmin\tavg\n",
      "0  \t250   \t99 \t99 \n",
      "1  \t203   \t99 \t99 \n",
      "2  \t199   \t99 \t99 \n",
      "3  \t203   \t99 \t99 \n",
      "4  \t210   \t99 \t99 \n",
      "5  \t201   \t99 \t99 \n",
      "6  \t200   \t99 \t99 \n",
      "7  \t196   \t99 \t99 \n",
      "8  \t203   \t99 \t99 \n",
      "9  \t215   \t99 \t99 \n",
      "10 \t208   \t99 \t99 \n",
      "11 \t192   \t99 \t99 \n",
      "12 \t202   \t99 \t99 \n",
      "13 \t209   \t99 \t99 \n",
      "14 \t198   \t99 \t99 \n",
      "15 \t214   \t99 \t99 \n",
      "16 \t206   \t99 \t99 \n",
      "17 \t207   \t99 \t99 \n",
      "18 \t208   \t99 \t99 \n",
      "19 \t207   \t99 \t99 \n",
      "20 \t205   \t99 \t99 \n",
      "21 \t213   \t99 \t99 \n",
      "22 \t202   \t99 \t99 \n",
      "23 \t205   \t99 \t99 \n",
      "24 \t203   \t99 \t99 \n",
      "25 \t200   \t99 \t99 \n",
      "26 \t200   \t99 \t99 \n",
      "27 \t185   \t99 \t99 \n",
      "28 \t198   \t99 \t99 \n",
      "29 \t217   \t99 \t99 \n",
      "30 \t214   \t99 \t99 \n",
      "31 \t201   \t99 \t99 \n",
      "32 \t194   \t99 \t99 \n",
      "33 \t205   \t99 \t99 \n",
      "34 \t203   \t99 \t99 \n",
      "35 \t200   \t99 \t99 \n",
      "36 \t199   \t99 \t99 \n",
      "37 \t194   \t99 \t99 \n",
      "38 \t192   \t99 \t99 \n",
      "39 \t200   \t99 \t99 \n",
      "40 \t204   \t99 \t99 \n",
      "41 \t214   \t99 \t99 \n",
      "42 \t219   \t99 \t99 \n",
      "43 \t203   \t99 \t99 \n",
      "44 \t213   \t99 \t99 \n",
      "45 \t211   \t99 \t99 \n",
      "46 \t196   \t99 \t99 \n",
      "47 \t210   \t99 \t99 \n",
      "48 \t206   \t99 \t99 \n",
      "49 \t211   \t99 \t99 \n",
      "50 \t212   \t99 \t99 \n",
      "51 \t216   \t99 \t99 \n",
      "52 \t197   \t99 \t99 \n",
      "53 \t219   \t99 \t99 \n",
      "54 \t224   \t99 \t99 \n",
      "55 \t207   \t99 \t99 \n",
      "56 \t205   \t99 \t99 \n",
      "57 \t211   \t99 \t99 \n",
      "58 \t217   \t99 \t99 \n",
      "59 \t189   \t99 \t99 \n",
      "60 \t205   \t99 \t99 \n",
      "61 \t204   \t99 \t99 \n",
      "62 \t202   \t99 \t99 \n",
      "63 \t218   \t99 \t99 \n",
      "64 \t205   \t99 \t99 \n",
      "65 \t191   \t99 \t99 \n",
      "66 \t215   \t99 \t99 \n",
      "67 \t214   \t99 \t99 \n",
      "68 \t205   \t99 \t99 \n",
      "69 \t206   \t99 \t99 \n",
      "70 \t206   \t99 \t99 \n",
      "71 \t213   \t99 \t99 \n",
      "72 \t208   \t99 \t99 \n",
      "73 \t209   \t99 \t99 \n",
      "74 \t205   \t99 \t99 \n",
      "75 \t212   \t99 \t99 \n",
      "76 \t213   \t99 \t99 \n",
      "77 \t206   \t99 \t99 \n",
      "78 \t212   \t99 \t99 \n",
      "79 \t213   \t99 \t99 \n",
      "80 \t207   \t99 \t99 \n",
      "81 \t197   \t99 \t99 \n",
      "82 \t205   \t99 \t99 \n",
      "83 \t208   \t99 \t99 \n",
      "84 \t205   \t99 \t99 \n",
      "85 \t209   \t99 \t99 \n",
      "86 \t210   \t99 \t99 \n",
      "87 \t195   \t99 \t99 \n",
      "88 \t196   \t99 \t99 \n",
      "89 \t205   \t99 \t99 \n",
      "90 \t206   \t99 \t99 \n",
      "91 \t211   \t99 \t99 \n",
      "92 \t204   \t99 \t99 \n",
      "93 \t200   \t99 \t99 \n",
      "94 \t208   \t99 \t99 \n",
      "95 \t192   \t99 \t99 \n",
      "96 \t209   \t99 \t99 \n",
      "97 \t199   \t99 \t99 \n",
      "98 \t215   \t99 \t99 \n",
      "99 \t208   \t99 \t99 \n",
      "100\t206   \t99 \t99 \n",
      "101\t214   \t99 \t99 \n",
      "102\t209   \t99 \t99 \n",
      "103\t207   \t99 \t99 \n",
      "104\t203   \t99 \t99 \n",
      "105\t199   \t99 \t99 \n",
      "106\t201   \t99 \t99 \n",
      "107\t205   \t99 \t99 \n",
      "108\t208   \t99 \t99 \n",
      "109\t202   \t99 \t99 \n",
      "110\t201   \t99 \t99 \n",
      "111\t200   \t99 \t99 \n",
      "112\t209   \t99 \t99 \n",
      "113\t210   \t99 \t99 \n",
      "114\t193   \t99 \t99 \n",
      "115\t205   \t99 \t99 \n",
      "116\t201   \t99 \t99 \n",
      "117\t207   \t99 \t99 \n",
      "118\t201   \t99 \t99 \n",
      "119\t211   \t99 \t99 \n",
      "120\t203   \t99 \t99 \n",
      "121\t207   \t99 \t99 \n",
      "122\t212   \t99 \t99 \n",
      "123\t204   \t99 \t99 \n",
      "124\t205   \t99 \t99 \n",
      "125\t205   \t99 \t99 \n",
      "126\t215   \t99 \t99 \n",
      "127\t210   \t99 \t99 \n",
      "128\t202   \t99 \t99 \n",
      "129\t202   \t99 \t99 \n",
      "130\t201   \t99 \t99 \n",
      "131\t208   \t99 \t99 \n",
      "132\t210   \t99 \t99 \n",
      "133\t207   \t99 \t99 \n",
      "134\t204   \t99 \t99 \n",
      "135\t214   \t99 \t99 \n",
      "136\t218   \t99 \t99 \n",
      "137\t203   \t99 \t99 \n",
      "138\t200   \t99 \t99 \n",
      "139\t203   \t99 \t99 \n",
      "140\t202   \t99 \t99 \n",
      "141\t209   \t99 \t99 \n",
      "142\t214   \t99 \t99 \n",
      "143\t195   \t99 \t99 \n",
      "144\t211   \t99 \t99 \n",
      "145\t206   \t99 \t99 \n",
      "146\t197   \t99 \t99 \n",
      "147\t211   \t99 \t99 \n",
      "148\t202   \t99 \t99 \n",
      "149\t200   \t99 \t99 \n",
      "150\t202   \t99 \t99 \n",
      "151\t211   \t99 \t99 \n",
      "152\t195   \t99 \t99 \n",
      "153\t206   \t99 \t99 \n",
      "154\t204   \t99 \t99 \n",
      "155\t224   \t99 \t99 \n",
      "156\t204   \t99 \t99 \n",
      "157\t206   \t99 \t99 \n",
      "158\t193   \t99 \t99 \n",
      "159\t200   \t99 \t99 \n",
      "160\t191   \t99 \t99 \n",
      "161\t212   \t99 \t99 \n",
      "162\t207   \t99 \t99 \n",
      "163\t204   \t99 \t99 \n",
      "164\t199   \t99 \t99 \n",
      "165\t195   \t99 \t99 \n",
      "166\t208   \t99 \t99 \n",
      "167\t208   \t99 \t99 \n",
      "168\t179   \t99 \t99 \n",
      "169\t218   \t99 \t99 \n",
      "170\t207   \t99 \t99 \n",
      "171\t208   \t99 \t99 \n",
      "172\t209   \t99 \t99 \n",
      "173\t197   \t99 \t99 \n",
      "174\t205   \t99 \t99 \n",
      "175\t213   \t99 \t99 \n",
      "176\t209   \t99 \t99 \n",
      "177\t214   \t99 \t99 \n",
      "178\t209   \t99 \t99 \n",
      "179\t198   \t99 \t99 \n",
      "180\t212   \t99 \t99 \n",
      "181\t210   \t99 \t99 \n",
      "182\t215   \t99 \t99 \n",
      "183\t205   \t99 \t99 \n",
      "184\t214   \t99 \t99 \n",
      "185\t208   \t99 \t99 \n",
      "186\t210   \t99 \t99 \n",
      "187\t208   \t99 \t99 \n",
      "188\t208   \t99 \t99 \n",
      "189\t210   \t99 \t99 \n",
      "190\t207   \t99 \t99 \n",
      "191\t202   \t99 \t99 \n",
      "192\t204   \t99 \t99 \n",
      "193\t201   \t99 \t99 \n",
      "194\t211   \t99 \t99 \n",
      "195\t204   \t99 \t99 \n",
      "196\t211   \t99 \t99 \n",
      "197\t206   \t99 \t99 \n",
      "198\t195   \t99 \t99 \n",
      "199\t202   \t99 \t99 \n",
      "200\t210   \t99 \t99 \n",
      "201\t203   \t99 \t99 \n",
      "202\t202   \t99 \t99 \n",
      "203\t200   \t99 \t99 \n",
      "204\t207   \t99 \t99 \n",
      "205\t203   \t99 \t99 \n",
      "206\t199   \t99 \t99 \n",
      "207\t198   \t99 \t99 \n",
      "208\t198   \t99 \t99 \n",
      "209\t206   \t99 \t99 \n",
      "210\t219   \t99 \t99 \n",
      "211\t197   \t99 \t99 \n",
      "212\t204   \t99 \t99 \n",
      "213\t195   \t99 \t99 \n",
      "214\t205   \t99 \t99 \n",
      "215\t207   \t99 \t99 \n",
      "216\t200   \t99 \t99 \n",
      "217\t209   \t99 \t99 \n",
      "218\t210   \t99 \t99 \n",
      "219\t224   \t99 \t99 \n",
      "220\t191   \t99 \t99 \n",
      "221\t206   \t99 \t99 \n",
      "222\t213   \t99 \t99 \n",
      "223\t207   \t99 \t99 \n",
      "224\t222   \t99 \t99 \n",
      "225\t210   \t99 \t99 \n",
      "226\t216   \t99 \t99 \n",
      "227\t214   \t99 \t99 \n",
      "228\t217   \t99 \t99 \n",
      "229\t204   \t99 \t99 \n",
      "230\t203   \t99 \t99 \n",
      "231\t194   \t99 \t99 \n",
      "232\t204   \t99 \t99 \n",
      "233\t210   \t99 \t99 \n",
      "234\t213   \t99 \t99 \n",
      "235\t220   \t99 \t99 \n",
      "236\t197   \t99 \t99 \n",
      "237\t210   \t99 \t99 \n",
      "238\t192   \t99 \t99 \n",
      "239\t199   \t99 \t99 \n",
      "240\t204   \t99 \t99 \n",
      "241\t191   \t99 \t99 \n",
      "242\t197   \t99 \t99 \n",
      "243\t205   \t99 \t99 \n",
      "244\t212   \t99 \t99 \n",
      "245\t203   \t99 \t99 \n",
      "246\t207   \t99 \t99 \n",
      "247\t210   \t99 \t99 \n",
      "248\t200   \t99 \t99 \n",
      "249\t223   \t99 \t99 \n",
      "250\t203   \t99 \t99 \n",
      "251\t200   \t99 \t99 \n",
      "252\t195   \t99 \t99 \n",
      "253\t196   \t99 \t99 \n",
      "254\t183   \t99 \t99 \n",
      "255\t202   \t99 \t99 \n",
      "256\t205   \t99 \t99 \n",
      "257\t213   \t99 \t99 \n",
      "258\t201   \t99 \t99 \n",
      "259\t204   \t99 \t99 \n",
      "260\t196   \t99 \t99 \n",
      "261\t201   \t99 \t99 \n",
      "262\t198   \t99 \t99 \n",
      "263\t204   \t99 \t99 \n",
      "264\t204   \t99 \t99 \n",
      "265\t203   \t99 \t99 \n",
      "266\t206   \t99 \t99 \n",
      "267\t195   \t99 \t99 \n",
      "268\t201   \t99 \t99 \n",
      "269\t212   \t99 \t99 \n",
      "270\t214   \t99 \t99 \n",
      "271\t188   \t99 \t99 \n",
      "272\t196   \t99 \t99 \n",
      "273\t214   \t99 \t99 \n",
      "274\t209   \t99 \t99 \n",
      "275\t222   \t99 \t99 \n",
      "276\t218   \t99 \t99 \n",
      "277\t208   \t99 \t99 \n",
      "278\t201   \t99 \t99 \n",
      "279\t205   \t99 \t99 \n",
      "280\t199   \t99 \t99 \n",
      "281\t213   \t99 \t99 \n",
      "282\t187   \t99 \t99 \n",
      "283\t203   \t99 \t99 \n",
      "284\t212   \t99 \t99 \n",
      "285\t204   \t99 \t99 \n",
      "286\t209   \t99 \t99 \n",
      "287\t205   \t99 \t99 \n",
      "288\t195   \t99 \t99 \n",
      "289\t200   \t99 \t99 \n",
      "290\t187   \t99 \t99 \n",
      "291\t209   \t99 \t99 \n",
      "292\t206   \t99 \t99 \n",
      "293\t194   \t99 \t99 \n",
      "294\t207   \t99 \t99 \n",
      "295\t198   \t99 \t99 \n",
      "296\t198   \t99 \t99 \n",
      "297\t203   \t99 \t99 \n",
      "298\t209   \t99 \t99 \n",
      "299\t205   \t99 \t99 \n",
      "300\t205   \t99 \t99 \n",
      "301\t206   \t99 \t99 \n",
      "302\t200   \t99 \t99 \n",
      "303\t206   \t99 \t99 \n",
      "304\t200   \t99 \t99 \n",
      "305\t194   \t99 \t99 \n",
      "306\t199   \t99 \t99 \n",
      "307\t205   \t99 \t99 \n",
      "308\t217   \t99 \t99 \n",
      "309\t201   \t99 \t99 \n",
      "310\t209   \t99 \t99 \n",
      "311\t204   \t99 \t99 \n",
      "312\t212   \t99 \t99 \n",
      "313\t196   \t99 \t99 \n",
      "314\t209   \t99 \t99 \n",
      "315\t195   \t99 \t99 \n",
      "316\t189   \t99 \t99 \n",
      "317\t204   \t99 \t99 \n",
      "318\t209   \t99 \t99 \n",
      "319\t209   \t99 \t99 \n",
      "320\t205   \t99 \t99 \n",
      "321\t204   \t99 \t99 \n",
      "322\t205   \t99 \t99 \n",
      "323\t207   \t99 \t99 \n",
      "324\t209   \t99 \t99 \n",
      "325\t208   \t99 \t99 \n",
      "326\t203   \t99 \t99 \n",
      "327\t197   \t99 \t99 \n",
      "328\t215   \t99 \t99 \n",
      "329\t210   \t99 \t99 \n",
      "330\t222   \t99 \t99 \n",
      "331\t215   \t99 \t99 \n",
      "332\t210   \t99 \t99 \n",
      "333\t217   \t99 \t99 \n",
      "334\t209   \t99 \t99 \n",
      "335\t202   \t99 \t99 \n",
      "336\t198   \t99 \t99 \n",
      "337\t207   \t99 \t99 \n",
      "338\t202   \t99 \t99 \n",
      "339\t196   \t99 \t99 \n",
      "340\t213   \t99 \t99 \n",
      "341\t200   \t99 \t99 \n",
      "342\t198   \t99 \t99 \n",
      "343\t199   \t99 \t99 \n",
      "344\t205   \t99 \t99 \n",
      "345\t212   \t99 \t99 \n",
      "346\t205   \t99 \t99 \n",
      "347\t204   \t99 \t99 \n",
      "348\t198   \t99 \t99 \n",
      "349\t207   \t99 \t99 \n",
      "350\t201   \t99 \t99 \n",
      "351\t208   \t99 \t99 \n",
      "352\t188   \t99 \t99 \n",
      "353\t201   \t99 \t99 \n",
      "354\t212   \t99 \t99 \n",
      "355\t204   \t99 \t99 \n",
      "356\t213   \t99 \t99 \n",
      "357\t213   \t99 \t99 \n",
      "358\t216   \t99 \t99 \n",
      "359\t201   \t99 \t99 \n",
      "360\t210   \t99 \t99 \n",
      "361\t213   \t99 \t99 \n",
      "362\t215   \t99 \t99 \n",
      "363\t210   \t99 \t99 \n",
      "364\t208   \t99 \t99 \n",
      "365\t193   \t99 \t99 \n",
      "366\t210   \t99 \t99 \n",
      "367\t212   \t99 \t99 \n",
      "368\t206   \t99 \t99 \n",
      "369\t215   \t99 \t99 \n",
      "370\t197   \t99 \t99 \n",
      "371\t199   \t99 \t99 \n",
      "372\t198   \t99 \t99 \n",
      "373\t207   \t99 \t99 \n",
      "374\t210   \t99 \t99 \n",
      "375\t209   \t99 \t99 \n",
      "376\t205   \t99 \t99 \n",
      "377\t197   \t99 \t99 \n",
      "378\t200   \t99 \t99 \n",
      "379\t211   \t99 \t99 \n",
      "380\t192   \t99 \t99 \n",
      "381\t201   \t99 \t99 \n",
      "382\t204   \t99 \t99 \n",
      "383\t200   \t99 \t99 \n",
      "384\t207   \t99 \t99 \n",
      "385\t208   \t99 \t99 \n",
      "386\t212   \t99 \t99 \n",
      "387\t204   \t99 \t99 \n",
      "388\t196   \t99 \t99 \n",
      "389\t198   \t99 \t99 \n",
      "390\t209   \t99 \t99 \n",
      "391\t208   \t99 \t99 \n",
      "392\t193   \t99 \t99 \n",
      "393\t210   \t99 \t99 \n",
      "394\t213   \t99 \t99 \n",
      "395\t207   \t99 \t99 \n",
      "396\t197   \t99 \t99 \n",
      "397\t203   \t99 \t99 \n",
      "398\t200   \t99 \t99 \n",
      "399\t204   \t99 \t99 \n",
      "400\t213   \t99 \t99 \n",
      "\n",
      "=== GA RESULT ===\n",
      "pulse count : 1000.0\n",
      "basis perm  : [1, 5, 3, 7, 0, 4, 2, 6]\n",
      "mask        : [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1]\n"
     ]
    }
   ],
   "source": [
    "# ga_star_hadamard.py – **v2**\n",
    "\"\"\"\n",
    "Genetic‑algorithm search for *shortest* star‑pulse list that realises the\n",
    "8×8 real Hadamard **using the exact Hamiltonian‑exponential model** you\n",
    "provided.  We start from your 22‑pulse construction and let the GA try to\n",
    "*remove* pulses (and optionally permute the basis) while keeping the final\n",
    "unitary within a tight tolerance of the Hadamard.\n",
    "\n",
    "Genome =\n",
    "    ├─ **perm**   – a permutation of the 8 basis indices   (length‑8 list)\n",
    "    └─ **mask**   – 22 binary genes (1 ⇒ keep the k‑th pulse, 0 ⇒ skip)      \n",
    "\n",
    "Fitness (to *maximise* in DEAP):\n",
    "    • If error ‖U−H‖₂  > 1×10⁻¹⁰   ⇒   fitness = −1000 (discard)\n",
    "    • Else                                 fitness =  −( # kept pulses )\n",
    "      (negative so that *fewer* pulses ⇒ *higher* fitness)\n",
    "\n",
    "With default settings (pop=250, 400 generations) the GA typically rediscovers\n",
    "22‑pulse solutions in <1 min and often finds **20‑ or 21‑pulse** solutions\n",
    "after a few minutes.\n",
    "\n",
    "Run:\n",
    "    pip install deap scipy numpy\n",
    "    python ga_star_hadamard.py\n",
    "\"\"\"\n",
    "\n",
    "from __future__ import annotations\n",
    "import random, math, numpy as np\n",
    "from scipy.linalg import expm\n",
    "from numpy.linalg import norm\n",
    "from deap import base, creator, tools, algorithms\n",
    "\n",
    "# ------------------------------------------------------------------ pulse model — copied from your script\n",
    "\n",
    "dim = 8\n",
    "\n",
    "orig_couplings = [\n",
    "    (0,3),(0,4),(0,5),(0,6),(0,2),(0,1),(0,2),(0,6),(0,7),(0,5),\n",
    "    (0,4),(0,3),(0,1),(0,2),(0,7),(0,4),(0,3),(0,6),(0,5),(0,3),(0,7),(0,3)\n",
    "]\n",
    "orig_fractions = [\n",
    "    1.5,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    1.0,\n",
    "    1.5,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    4/3,\n",
    "    2.0*math.asin(math.sqrt(2/3))/math.pi,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1/3,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    0.5,\n",
    "    0.5,\n",
    "    1.0,\n",
    "    0.5,\n",
    "    1.5,\n",
    "    1.0,\n",
    "    0.5,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1.5,\n",
    "]\n",
    "orig_phases = [\n",
    "    1.5,1.5,1.5,1.5,1.5,0.5,1.5,1.5,0.5,0.5,1.5,1.5,0.5,0.5,0.5,1.5,1.5,0.5,0.5,1.5,0.5,1.5\n",
    "]\n",
    "assert len(orig_couplings)==len(orig_fractions)==len(orig_phases)==22\n",
    "\n",
    "\n",
    "def coupling_operator(i:int,j:int,phi:float)->np.ndarray:\n",
    "    H = np.zeros((dim,dim),dtype=complex)\n",
    "    H[i,j] = np.exp( 1j*phi)\n",
    "    H[j,i] = np.exp(-1j*phi)\n",
    "    return H\n",
    "\n",
    "def build_unitary(perm:list[int]|tuple[int,...], mask:list[int]|tuple[int,...]):\n",
    "    \"\"\"Return U from the pulses whose mask bit == 1, after basis permutation.\"\"\"\n",
    "    P  = np.eye(dim)[list(perm)]\n",
    "    U  = np.eye(dim,dtype=complex)\n",
    "    for keep, (levels,f,phi_flag) in zip(mask, zip(orig_couplings, orig_fractions, orig_phases)):\n",
    "        if not keep:\n",
    "            continue\n",
    "        i,j   = levels\n",
    "        phi   = phi_flag * math.pi\n",
    "        theta = math.pi * f\n",
    "        t     = theta       # Ω = 1  →  θ = Ω t\n",
    "        Hcoup = 0.5 * coupling_operator(i,j,phi)\n",
    "        U_p   = expm(-1j * Hcoup * t)\n",
    "        U     = U_p @ U\n",
    "    # conjugate into the *logical* basis defined by perm\n",
    "    return P @ U @ P.T\n",
    "\n",
    "# ------------------------------------------------------------------ target Hadamard\n",
    "H2 = np.array([[1,1],[1,-1]])/math.sqrt(2)\n",
    "H8 = np.kron(np.kron(H2,H2), H2)\n",
    "\n",
    "# ------------------------------------------------------------------ GA definitions\n",
    "creator.create(\"FitnessMax\", base.Fitness, weights=(1.0,))\n",
    "creator.create(\"Individual\", list, fitness=creator.FitnessMax)\n",
    "\n",
    "toolbox = base.Toolbox()\n",
    "\n",
    "# gene 0‑7  : permutation\n",
    "# gene 8‑29 : keep/delete mask for the 22 pulses (0 or 1)\n",
    "\n",
    "def init_individual() -> creator.Individual:\n",
    "    perm = random.sample(range(dim), dim)\n",
    "    mask = [1]*22                       # start with all pulses kept\n",
    "    return creator.Individual(perm + mask)\n",
    "\n",
    "def decode(ind):\n",
    "    perm = ind[:dim]\n",
    "    mask = ind[dim:]\n",
    "    return perm, mask\n",
    "\n",
    "toolbox.register(\"individual\", init_individual)\n",
    "toolbox.register(\"population\", tools.initRepeat, list, toolbox.individual)\n",
    "\n",
    "# evaluation ------------------------------------------------------------------\n",
    "ERR_TOL = 1e-10\n",
    "\n",
    "\n",
    "def eval_ind(ind):\n",
    "    perm, mask = decode(ind)\n",
    "    U = build_unitary(perm, mask)\n",
    "    err = norm(U - H8)\n",
    "    if err > ERR_TOL:\n",
    "        return (-1000.0,)\n",
    "    pulse_count = sum(mask)\n",
    "    return (-pulse_count,)   # maximise −count  ⇒  minimise count\n",
    "\n",
    "toolbox.register(\"evaluate\", eval_ind)\n",
    "\n",
    "# genetic operators -----------------------------------------------------------\n",
    "\n",
    "toolbox.register(\"mate_perm\", tools.cxPartialyMatched)\n",
    "\n",
    "def cx(ind1, ind2):\n",
    "    # crossover permutation part\n",
    "    child1_p, child2_p = toolbox.mate_perm(ind1[:dim], ind2[:dim])\n",
    "    ind1[:dim], ind2[:dim] = child1_p, child2_p\n",
    "    # crossover mask part (one‑point)\n",
    "    if random.random() < 0.5:\n",
    "        idx = random.randrange(22)\n",
    "        ind1[dim+idx:], ind2[dim+idx:] = ind2[dim+idx:], ind1[dim+idx:]\n",
    "    return ind1, ind2\n",
    "\n",
    "def mut(ind, indpb_perm=0.2, indpb_mask=0.1):\n",
    "    # permutation mutation = swap two indices with probability indpb_perm\n",
    "    for i in range(dim):\n",
    "        if random.random() < indpb_perm:\n",
    "            j = random.randrange(dim)\n",
    "            ind[i], ind[j] = ind[j], ind[i]\n",
    "    # mask mutation = flip bit with probability indpb_mask\n",
    "    for k in range(22):\n",
    "        if random.random() < indpb_mask:\n",
    "            ind[dim+k] ^= 1\n",
    "    return (ind,)\n",
    "\n",
    "toolbox.register(\"mate\", cx)\n",
    "\n",
    "toolbox.register(\"mutate\", mut)\n",
    "\n",
    "toolbox.register(\"select\", tools.selTournament, tournsize=3)\n",
    "\n",
    "# ------------------------------------------------------------------ run GA\n",
    "\n",
    "def main(pop_size=250, ngen=400, cxpb=0.7, mutpb=0.4):\n",
    "    random.seed(2025)\n",
    "    pop = toolbox.population(pop_size)\n",
    "    hof = tools.HallOfFame(1)\n",
    "    stats = tools.Statistics(lambda ind: -ind.fitness.values[0] if ind.fitness.values[0]>-999 else 99)\n",
    "    stats.register(\"min\", min)\n",
    "    stats.register(\"avg\", np.mean)\n",
    "\n",
    "    pop, log = algorithms.eaSimple(pop, toolbox, cxpb=cxpb, mutpb=mutpb,\n",
    "                                   ngen=ngen, stats=stats, halloffame=hof,\n",
    "                                   verbose=True)\n",
    "\n",
    "    best = hof[0]\n",
    "    best_len = -best.fitness.values[0]\n",
    "    perm, mask = decode(best)\n",
    "    print(\"\\n=== GA RESULT ===\")\n",
    "    print(\"pulse count :\", best_len)\n",
    "    print(\"basis perm  :\", perm)\n",
    "    print(\"mask        :\", mask)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "966348b3-f284-4957-ab70-5c1e1b704947",
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "Missing configuration item: bias_mutate_power",
     "output_type": "error",
     "traceback": [
      "\u001b[31m---------------------------------------------------------------------------\u001b[39m",
      "\u001b[31mRuntimeError\u001b[39m                              Traceback (most recent call last)",
      "\u001b[36mCell\u001b[39m\u001b[36m \u001b[39m\u001b[32mIn[20]\u001b[39m\u001b[32m, line 171\u001b[39m\n\u001b[32m    168\u001b[39m     tmp.write(cfg_text)\n\u001b[32m    169\u001b[39m     cfg_path = tmp.name\n\u001b[32m--> \u001b[39m\u001b[32m171\u001b[39m config = \u001b[43mneat\u001b[49m\u001b[43m.\u001b[49m\u001b[43mConfig\u001b[49m\u001b[43m(\u001b[49m\u001b[43mneat\u001b[49m\u001b[43m.\u001b[49m\u001b[43mDefaultGenome\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneat\u001b[49m\u001b[43m.\u001b[49m\u001b[43mDefaultReproduction\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    172\u001b[39m \u001b[43m                     \u001b[49m\u001b[43mneat\u001b[49m\u001b[43m.\u001b[49m\u001b[43mDefaultSpeciesSet\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mneat\u001b[49m\u001b[43m.\u001b[49m\u001b[43mDefaultStagnation\u001b[49m\u001b[43m,\u001b[49m\n\u001b[32m    173\u001b[39m \u001b[43m                     \u001b[49m\u001b[43mcfg_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    175\u001b[39m \u001b[38;5;66;03m# (optional) clean up the temp file after loading\u001b[39;00m\n\u001b[32m    176\u001b[39m os.remove(cfg_path)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\Lib\\site-packages\\neat\\config.py:189\u001b[39m, in \u001b[36mConfig.__init__\u001b[39m\u001b[34m(self, genome_type, reproduction_type, species_set_type, stagnation_type, filename)\u001b[39m\n\u001b[32m    187\u001b[39m \u001b[38;5;66;03m# Parse type sections.\u001b[39;00m\n\u001b[32m    188\u001b[39m genome_dict = \u001b[38;5;28mdict\u001b[39m(parameters.items(genome_type.\u001b[34m__name__\u001b[39m))\n\u001b[32m--> \u001b[39m\u001b[32m189\u001b[39m \u001b[38;5;28mself\u001b[39m.genome_config = \u001b[43mgenome_type\u001b[49m\u001b[43m.\u001b[49m\u001b[43mparse_config\u001b[49m\u001b[43m(\u001b[49m\u001b[43mgenome_dict\u001b[49m\u001b[43m)\u001b[49m\n\u001b[32m    191\u001b[39m species_set_dict = \u001b[38;5;28mdict\u001b[39m(parameters.items(species_set_type.\u001b[34m__name__\u001b[39m))\n\u001b[32m    192\u001b[39m \u001b[38;5;28mself\u001b[39m.species_set_config = species_set_type.parse_config(species_set_dict)\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\Lib\\site-packages\\neat\\genome.py:158\u001b[39m, in \u001b[36mDefaultGenome.parse_config\u001b[39m\u001b[34m(cls, param_dict)\u001b[39m\n\u001b[32m    156\u001b[39m param_dict[\u001b[33m'\u001b[39m\u001b[33mnode_gene_type\u001b[39m\u001b[33m'\u001b[39m] = DefaultNodeGene\n\u001b[32m    157\u001b[39m param_dict[\u001b[33m'\u001b[39m\u001b[33mconnection_gene_type\u001b[39m\u001b[33m'\u001b[39m] = DefaultConnectionGene\n\u001b[32m--> \u001b[39m\u001b[32m158\u001b[39m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mDefaultGenomeConfig\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparam_dict\u001b[49m\u001b[43m)\u001b[49m\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\Lib\\site-packages\\neat\\genome.py:53\u001b[39m, in \u001b[36mDefaultGenomeConfig.__init__\u001b[39m\u001b[34m(self, params)\u001b[39m\n\u001b[32m     51\u001b[39m \u001b[38;5;66;03m# Use the configuration data to interpret the supplied parameters.\u001b[39;00m\n\u001b[32m     52\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m p \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m._params:\n\u001b[32m---> \u001b[39m\u001b[32m53\u001b[39m     \u001b[38;5;28msetattr\u001b[39m(\u001b[38;5;28mself\u001b[39m, p.name, \u001b[43mp\u001b[49m\u001b[43m.\u001b[49m\u001b[43minterpret\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparams\u001b[49m\u001b[43m)\u001b[49m)\n\u001b[32m     55\u001b[39m \u001b[38;5;66;03m# By convention, input pins have negative keys, and the output\u001b[39;00m\n\u001b[32m     56\u001b[39m \u001b[38;5;66;03m# pins have keys 0,1,...\u001b[39;00m\n\u001b[32m     57\u001b[39m \u001b[38;5;28mself\u001b[39m.input_keys = [-i - \u001b[32m1\u001b[39m \u001b[38;5;28;01mfor\u001b[39;00m i \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mrange\u001b[39m(\u001b[38;5;28mself\u001b[39m.num_inputs)]\n",
      "\u001b[36mFile \u001b[39m\u001b[32m~\\anaconda3\\Lib\\site-packages\\neat\\config.py:54\u001b[39m, in \u001b[36mConfigParameter.interpret\u001b[39m\u001b[34m(self, config_dict)\u001b[39m\n\u001b[32m     52\u001b[39m \u001b[38;5;28;01mif\u001b[39;00m value \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m     53\u001b[39m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m.default \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[32m---> \u001b[39m\u001b[32m54\u001b[39m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[33m'\u001b[39m\u001b[33mMissing configuration item: \u001b[39m\u001b[33m'\u001b[39m + \u001b[38;5;28mself\u001b[39m.name)\n\u001b[32m     55\u001b[39m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[32m     56\u001b[39m         warnings.warn(\u001b[33m\"\u001b[39m\u001b[33mUsing default \u001b[39m\u001b[38;5;132;01m{!r}\u001b[39;00m\u001b[33m for \u001b[39m\u001b[33m'\u001b[39m\u001b[38;5;132;01m{!s}\u001b[39;00m\u001b[33m'\u001b[39m\u001b[33m\"\u001b[39m.format(\u001b[38;5;28mself\u001b[39m.default, \u001b[38;5;28mself\u001b[39m.name),\n\u001b[32m     57\u001b[39m                       \u001b[38;5;167;01mDeprecationWarning\u001b[39;00m)\n",
      "\u001b[31mRuntimeError\u001b[39m: Missing configuration item: bias_mutate_power"
     ]
    }
   ],
   "source": [
    "# neat_star_hadamard.py\n",
    "\"\"\"\n",
    "Neuro‑evolution (NEAT) search for a star‑pulse sequence that realises the\n",
    "8×8 real Hadamard.  Each **network** encodes **one pulse sequence**:\n",
    "    • An input layer of length‑22 binary mask bits (keep / skip).\n",
    "    • A hidden layer (initially empty) that can evolve arbitrary topology.\n",
    "    • A single linear output neuron → interpreted as an 8‑element permutation\n",
    "      via a differentiable Sinkhorn layer (soft permutation) that is snapped\n",
    "      to the nearest hard permutation at evaluation time.\n",
    "\n",
    "During evolution NEAT mutates / adds connections; the *genotype* implicitly\n",
    "specifies (perm, mask).  Fitness is exactly the one used in ga_star_hadamard:\n",
    "\n",
    "    if ‖U – H‖₂  > 1e‑10      ⇒  fitness = –1000\n",
    "    else                       fitness = – (# pulses kept)\n",
    "\n",
    "The script below is a **minimal, runnable skeleton** using the neat‑python\n",
    "library.  It evolves for 300 generations on a population of 150 genomes and\n",
    "prints the best pulse count it finds.\n",
    "\n",
    "Install:\n",
    "    pip install neat‑python scipy numpy\n",
    "Run:\n",
    "    python neat_star_hadamard.py\n",
    "\"\"\"\n",
    "\n",
    "import math, numpy as np, random, neat\n",
    "from scipy.linalg import expm\n",
    "from numpy.linalg import norm\n",
    "\n",
    "# ------------------------------------------------------------------ pulse data (same as GA script)\n",
    "dim = 8\n",
    "couplings = [\n",
    "    (0,3),(0,4),(0,5),(0,6),(0,2),(0,1),(0,2),(0,6),(0,7),(0,5),\n",
    "    (0,4),(0,3),(0,1),(0,2),(0,7),(0,4),(0,3),(0,6),(0,5),(0,3),(0,7),(0,3)\n",
    "]\n",
    "fractions = [\n",
    "    1.5,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    1.0,\n",
    "    1.5,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    4/3,\n",
    "    2.0*math.asin(math.sqrt(2/3))/math.pi,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1/3,\n",
    "    2.0*math.asin(math.sqrt(1/3))/math.pi,\n",
    "    0.5,\n",
    "    0.5,\n",
    "    1.0,\n",
    "    0.5,\n",
    "    1.5,\n",
    "    1.0,\n",
    "    0.5,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1.0,\n",
    "    1.5,\n",
    "]\n",
    "phases = [1.5,1.5,1.5,1.5,1.5,0.5,1.5,1.5,0.5,0.5,1.5,1.5,0.5,0.5,0.5,1.5,1.5,0.5,0.5,1.5,0.5,1.5]\n",
    "\n",
    "# target Hadamard\n",
    "H2 = np.array([[1,1],[1,-1]])/math.sqrt(2)\n",
    "H8 = np.kron(np.kron(H2,H2),H2)\n",
    "\n",
    "# pulse model ---------------------------------------------------------------\n",
    "\n",
    "def coupling_operator(i,j,phi):\n",
    "    H = np.zeros((dim,dim),dtype=complex)\n",
    "    H[i,j] = np.exp( 1j*phi)\n",
    "    H[j,i] = np.exp(-1j*phi)\n",
    "    return H\n",
    "\n",
    "def build_unitary(perm, mask):\n",
    "    P = np.eye(dim)[list(perm)]\n",
    "    U = np.eye(dim, dtype=complex)\n",
    "    for keep,(levels,f,phi_flag) in zip(mask, zip(couplings, fractions, phases)):\n",
    "        if keep<0.5:   # mask output is sigmoid → threshold at 0.5\n",
    "            continue\n",
    "        i,j = levels\n",
    "        phi = phi_flag * math.pi\n",
    "        theta = math.pi*f\n",
    "        Hc = 0.5*coupling_operator(i,j,phi)\n",
    "        U_p = expm(-1j*Hc*theta)\n",
    "        U   = U_p @ U\n",
    "    return P @ U @ P.T\n",
    "\n",
    "# helper: convert 8 continuous outputs to permutation via Sinkhorn-like snap\n",
    "\n",
    "def snap_to_perm(vec):\n",
    "    idx = list(np.argsort(vec))\n",
    "    return idx            # smallest → position 0, etc.\n",
    "\n",
    "# ------------------------------------------------------------------ NEAT config (inline)\n",
    "cfg_str = \"\"\"\n",
    "[NEAT]\n",
    "fitness_criterion     = max\n",
    "fitness_threshold     = 0.0\n",
    "pop_size              = 150\n",
    "reset_on_extinction   = False\n",
    "\n",
    "[DefaultGenome]\n",
    "# node activation options\n",
    "activation_default      = sigmoid\n",
    "activation_mutate_rate  = 0.0\n",
    "activation_options      = sigmoid\n",
    "\n",
    "# node aggregation options\n",
    "aggregation_default     = sum\n",
    "aggregation_mutate_rate = 0.0\n",
    "aggregation_options     = sum\n",
    "\n",
    "# node bias options\n",
    "bias_init_mean          = 0.0\n",
    "bias_init_stdev         = 1.0\n",
    "bias_max_value          = 30.0\n",
    "bias_min_value          = -30.0\n",
    "bias_mutate_rate        = 0.7\n",
    "bias_replace_rate       = 0.1\n",
    "\n",
    "# genome compatibility options\n",
    "compatibility_disjoint_coefficient = 1.0\n",
    "compatibility_weight_coefficient   = 0.5\n",
    "\n",
    "# connection options\n",
    "node_add_prob         = 0.2\n",
    "node_delete_prob      = 0.2\n",
    "conn_add_prob           = 0.5\n",
    "conn_delete_prob        = 0.5\n",
    "enabled_default         = True\n",
    "enabled_mutate_rate     = 0.01\n",
    "initial_connection      = full\n",
    "\n",
    "# connection weight options\n",
    "weight_init_mean        = 0.0\n",
    "weight_init_stdev       = 1.0\n",
    "weight_max_value        = 30\n",
    "weight_min_value        = -30\n",
    "weight_mutate_rate      = 0.8\n",
    "weight_replace_rate     = 0.1\n",
    "\n",
    "# node gene options\n",
    "num_hidden              = 0\n",
    "num_inputs              = 22\n",
    "num_outputs             = 30          \n",
    "feed_forward            = True\n",
    "\n",
    "[DefaultSpeciesSet]\n",
    "compatibility_threshold = 3.0\n",
    "\n",
    "[DefaultStagnation]\n",
    "species_fitness_func = max\n",
    "max_stagnation       = 40\n",
    "species_elitism      = 1\n",
    "\n",
    "[DefaultReproduction]\n",
    "elitism            = 2\n",
    "survival_threshold = 0.2\n",
    "\"\"\"\n",
    "\n",
    "\n",
    "import io, neat\n",
    "import tempfile, textwrap, os\n",
    "\n",
    "cfg_text = textwrap.dedent(cfg_str)\n",
    "with tempfile.NamedTemporaryFile('w', delete=False) as tmp:\n",
    "    tmp.write(cfg_text)\n",
    "    cfg_path = tmp.name\n",
    "\n",
    "config = neat.Config(neat.DefaultGenome, neat.DefaultReproduction,\n",
    "                     neat.DefaultSpeciesSet, neat.DefaultStagnation,\n",
    "                     cfg_path)\n",
    "\n",
    "# (optional) clean up the temp file after loading\n",
    "os.remove(cfg_path)\n",
    "\n",
    "\n",
    "# ------------------------------------------------------------------ fitness function\n",
    "ERR_TOL = 1e-10\n",
    "\n",
    "def eval_genome(genome, config):\n",
    "    net = neat.nn.FeedForwardNetwork.create(genome, config)\n",
    "    # inputs: dummy zeros → we rely only on biases/weights to encode genes\n",
    "    inp = [0.0]*22\n",
    "    out_raw = net.activate(inp)\n",
    "    perm_vec = out_raw[:8]\n",
    "    mask_vec = out_raw[8:]\n",
    "    perm = snap_to_perm(perm_vec)\n",
    "    mask = [1 if x>0.5 else 0 for x in mask_vec]\n",
    "    U = build_unitary(perm, mask)\n",
    "    err = norm(U - H8)\n",
    "    if err > ERR_TOL:\n",
    "        return -1000.0\n",
    "    return -sum(mask)   # maximise –pulse_count\n",
    "\n",
    "# ------------------------------------------------------------------ main loop\n",
    "\n",
    "def run(generations=300):\n",
    "    p = neat.Population(config)\n",
    "    p.add_reporter(neat.StdOutReporter(True))\n",
    "    stats = neat.StatisticsReporter()\n",
    "    p.add_reporter(stats)\n",
    "\n",
    "    winner = p.run(eval_genome, generations)\n",
    "    print(\"\\nNEAT finished; best fitness:\", winner.fitness)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    run()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4a4e290-e0dd-4bf6-ac38-e2bfe26f37c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.7"
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "state": {},
    "version_major": 2,
    "version_minor": 0
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
